feat: Add Object Storage service

Geun-Oh opened the pull request:
## Summary

Add Object Storage service(including resource, data source, tests, guide docs and example codes)

## Description

1. objectstorage_bucket_acl.go

2. objectstorage_bucket_data_source.go

3. objectstorage_bucket.go

4. objectstorage_object_acl.go

5. objectstorage_object_data_source.go

6. objectstorage_object.go

7. objectstorage_object_copy.go

8. resource & datasource test go files

9. resource & datasource docs(.md) files

## Related Issue

- #425 

## Error & Additional

- This resource is implemented with aws s3 compatible sdk.

Geun-Oh committed:
chore: Install AWS SDK for Object Storage API

Geun-Oh committed:
feat: Add S3 Client configuration

Geun-Oh committed:
feat: Add Object Storage Resources

Geun-Oh committed:
feat: Add objectstorage_bucket resource

Geun-Oh committed:
feat: Add objectstorage_bucket_acl resource

Geun-Oh committed:
feat: Add objectstorage_object resource

Geun-Oh committed:
feat: Add objectstorage_object_acl resource

Geun-Oh committed:
fix: add "_bucket" suffix to bucket resource

Geun-Oh committed:
feat: Add examples of object storage

Geun-Oh committed:
docs: Add initial documentation of object storage

Geun-Oh committed:
feat: add type casting function for nested listValue

Geun-Oh committed:
feat: Add BucketRegion to bucket chema

Geun-Oh committed:
fix: Add detailed schema of grants & type casting method

Geun-Oh committed:
chore: Add error exception

Geun-Oh committed:
feat: Add regex validator for resource id

Geun-Oh committed:
chore: remove redundant EOL

Geun-Oh committed:
fix: Remove BucketName, BucketRegion from schema(doesn't provided)

Geun-Oh committed:
refactor: Add TrimForIDParsing util

Geun-Oh committed:
feat: Add temporary file creation logic

Geun-Oh committed:
feat: Add aclOptions for testing

Geun-Oh committed:
fix: Removed Unavailable object acl

Geun-Oh committed:
feat: Add Bucket data source

Geun-Oh committed:
fix: Add new bucket creation for independant testing

Geun-Oh committed:
fix: aws.String => ncloud.String

Geun-Oh committed:
chore: fix typo

Geun-Oh committed:
chore: change string value with constants

Geun-Oh committed:
refactor: expand TrimForParsing to parse vairous strings (not only ID)

Geun-Oh committed:
feat: use region info from initial configuration

Geun-Oh committed:
refactor: Set bucket id same as bucket name

Geun-Oh committed:
fix: Add detailed string validator with regex for bucket_name

Geun-Oh committed:
feat: Add NewObjectDataSource data source

Geun-Oh committed:
refactor: Move ObjectIDParser to object.go

Geun-Oh committed:
refactor: Set Object Resource ID with new format

Geun-Oh committed:
feat: set independent object testing

Geun-Oh committed:
feat: Add Owner info & Creation Date to bucket data source

Geun-Oh committed:
feat: Add OwnerID & OwnerDisplayname at object_acl attribute

Geun-Oh committed:
fix: set bucket_id & object_id w/o NRN

Geun-Oh committed:
docs: Add documentation of object storage resources

Geun-Oh committed:
feat: Change format - make resource id with '/' in object storage - object

Geun-Oh committed:
feat: Add objectstorage_object_copy

Geun-Oh committed:
fix: Change to do not use pointer method

Geun-Oh committed:
docs: Add data source documentation of bucket

Geun-Oh committed:
docs: Add object_storage_object.md in data-source

Geun-Oh committed:
docs: Add object_storage_object_copy.md

Geun-Oh committed:
feat: Add CreationDate attribute to bucket

Geun-Oh committed:
fix: fix logic to check whether output is nil

Geun-Oh committed:
fix: Add format of creationDate

Geun-Oh committed:
feat: Delete panic from object_copy Update

Geun-Oh committed:
feat: Remove endpoint from Config struct

Geun-Oh committed:
fix: Remove Optional keyword from creation_date(computed)

Geun-Oh committed:
refactor: set bucket unique value from bucket_id to bucket_name

Geun-Oh committed:
fix: Remove unused headbucket

Geun-Oh committed:
feat: Add bucket ID (since it uses id for default)

Geun-Oh committed:
feat: Add attributes based on HeadObjectOutput

Geun-Oh committed:
feat: Add endpoint abstract logic

Geun-Oh committed:
refactor: Move post fetch logic into refreshFromOutput

Geun-Oh committed:
feat: Add genEndpointWithCode (region, site)

Geun-Oh committed:
feat: Add optional endpoint configuration

Geun-Oh committed:
feat: Add region formatting logic

Geun-Oh committed:
feat: Add Optional variables compatible with aws sdk in object-kind resource

Geun-Oh committed:
docs: Update docs with additional attributes

Geun-Oh committed:
fix: Remove Object Encryption logic entirely

Geun-Oh committed:
feat: Add site variable in example tf code

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,281 @@
+	}
+
+	output, err := o.config.Client.ObjectStorage.ListBuckets(ctx, &s3.ListBucketsInput{})
+	if err != nil {
```
무시해도 무방한 에러가 아니라면, 에러 발생시 `resp.Diagnostics.AddError()` 를 사용하여 에러 처리해주면 좋겠습니다.

Geun-Oh replied:
완전 잘못 처리하고 있었군요...확인했습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket.go:
```diff
@@ -0,0 +1,281 @@
+			_, err := o.config.Client.ObjectStorage.HeadBucket(ctx, &s3.HeadBucketInput{
+				Bucket: plan.BucketName.ValueStringPointer(),
+			})
+			if err != nil {
```
(에러처리)

wonchulee commented on internal/service/objectstorage/objectstorage_object_acl.go:
```diff
@@ -0,0 +1,341 @@
+	}
+	if output == nil {
+		return
+	}
```
GetObjectAcl() 실패시 diagnostic를 사용한 에러처리 추가해주시면 좋겠습니다.

Geun-Oh committed:
chore: Upgrade Go version to v1.21

chore: Upgrade Go version to v1.21

chore: upgrade Go to v1.21

ci: Set go version 1.21 in GitHub Actions

ci: Upgrade actions version (checkout, setup-go, ci-lint)

ci: Upgrade golangci-lint v1.59

wonchulee commented on internal/service/objectstorage/objectstorage_object_acl.go:
```diff
@@ -0,0 +1,341 @@
+	if diag.HasError() {
+		fmt.Printf("Error Occured with parsing Grants to ListValue")
+		return
+	}
```
에러 발생시 printf 대신 diagnostic을 반환해 resp.Diagnostic을 통해 framework에서 에러 처리 및 로깅하면 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_object_copy.go:
```diff
@@ -0,0 +1,373 @@
+
+	output, err := o.config.Client.ObjectStorage.CopyObject(ctx, reqParams)
+	if err != nil {
+		resp.Diagnostics.AddError("COPYING ERROR", err.Error())
```
여기서 return 이 누락된 것 같습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_object_copy.go:
```diff
@@ -0,0 +1,373 @@
+							regexp.MustCompile(`^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$|.+)`),
+							"Bucket name cannot be formatted as an IP address.",
+						),
+					),
```
사소한 부분이지만, 여러 리소스에 걸쳐 동일한 조건의 validator가 사용되네요, 따로 정의하고 재사용하는 형태로 만들어 주셔도 좋겠습니다.

Geun-Oh replied:
해당 부분 `objectstorage_bucket.go` 내 `BucketNameValidator()` 를 통해 재사용하도록 수정하였습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_object_copy.go:
```diff
@@ -0,0 +1,373 @@
+				},
+				Description: "(Required) Name of the object once it is in the bucket",
+			},
+			"source": schema.StringAttribute{
+				Required:    true,
+				Description: "(Required) Path of the object",
+			},
```
source 값 업데이트시 동작 구현이 누락된 것으로 보입니다. 리소스 재생성 또는 Update() 동작 구현이 추가될 필요가 있습니다.

Geun-Oh replied:
source 변경사항 기반으로 새로운 파일을 업데이트 할 수 있도록 해 두었고, 파일에 종속된 속성값들 또한 비교하여 업데이트되도록 코드 추가해두었습니다 :)

wonchulee commented on internal/service/objectstorage/objectstorage_object_data_source.go:
```diff
@@ -0,0 +1,209 @@
+		Bucket: ncloud.String(bucketName),
+		Key:    ncloud.String(key),
+	})
+	if err != nil {
```
`resp.Diagnostics.AddError()`를 사용한 에러 처리 추가해주시면 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_object_data_source.go:
```diff
@@ -0,0 +1,209 @@
+	}
+
+	bodyBytes, err := io.ReadAll(output.Body)
+	if err != nil {
```
(에러처리)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_data_source.go:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	output, err := b.config.Client.ObjectStorage.ListBuckets(ctx, &s3.ListBucketsInput{})
+	if err != nil {
```
`resp.Diagnostics.AddError()`를 사용한 에러 처리 추가해주시면 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_data_source.go:
```diff
@@ -0,0 +1,137 @@
+			_, err := b.config.Client.ObjectStorage.HeadBucket(ctx, &s3.HeadBucketInput{
+				Bucket: data.BucketName.ValueStringPointer(),
+			})
+			if err != nil {
```
(에러처리)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,345 @@
+
+	response, err := b.config.Client.ObjectStorage.PutBucketAcl(ctx, reqParams)
+	if err != nil {
+		resp.Diagnostics.AddError("CREATING ERROR", err.Error())
```
여기서 return 이 누락된 것 같습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,345 @@
+	}
+	if output == nil {
+		return
+	}
```
다른 리소스에 코멘트 남긴 것과 동일하게 diagnostic을 통한 에러처리 추가해주시면 좋겠습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,345 @@
+	if diag.HasError() {
+		fmt.Printf("Error Occured with parsing Grants to ListValue")
+		return
+	}
```
(에러처리)

wonchulee commented on internal/service/objectstorage/objectstorage_object.go:
```diff
@@ -0,0 +1,400 @@
+
+	output, err := o.config.Client.ObjectStorage.PutObject(ctx, reqParams)
+	if err != nil {
+		resp.Diagnostics.AddError("CREATING ERROR", err.Error())
```
여기서 return이 누락된 것 같습니다.

wonchulee commented on internal/service/objectstorage/objectstorage_object.go:
```diff
@@ -0,0 +1,400 @@
+				},
+				Description: "(Required) Name of the object once it is in the bucket",
+			},
+			"source": schema.StringAttribute{
+				Required:    true,
+				Description: "(Required) Path of the object",
+			},
```
(object copy 리소스의 `source` attribute에 남긴 코멘트와 동일한 부분입니다.)

wonchulee commented on internal/service/objectstorage/objectstorage_object.go:
```diff
@@ -0,0 +1,400 @@
+		Bucket: o.Bucket.ValueStringPointer(),
+		Key:    o.Key.ValueStringPointer(),
+	})
+	if err != nil {
```
(에러처리)

wonchulee commented on internal/service/objectstorage/objectstorage_object_acl.go:
```diff
@@ -0,0 +1,341 @@
+
+	response, err := o.config.Client.ObjectStorage.PutObjectAcl(ctx, reqParams)
+	if err != nil {
+		resp.Diagnostics.AddError("CREATING ERROR", err.Error())
```
(return 누락된 것 같아요)

wonchulee commented on internal/service/objectstorage/objectstorage_object_copy.go:
```diff
@@ -0,0 +1,373 @@
+		Key:    o.Key.ValueStringPointer(),
+	})
+	if err != nil {
+		return
```
(에러처리)

wonchulee commented on internal/conn/config.go:
```diff
@@ -113,5 +117,6 @@ type ProviderConfig struct {
 	SupportVPC bool
 	RegionCode string
 	RegionNo   string
+	Endpoint   string
```
사소한 부분이지만, Endpoint는 objectstorage의 s3 client에만 적용되는 값으로 보여, ProviderConfig에서 제거하고 client가 소유한 값으로 변경해주시면 좋겠습니다.

Geun-Oh replied:
환경변수에서 값을 가져오는 작업들이 모두 provider.go 에서 이루어지고 있는 부분이 있어 해당 로직을 provider.go 에 둔 바 있습니다.

인자로 가져가더라도 환경변수를 불러오는 역할을 한 파일에 집중시키는 것은 어떤지 의견 드립니다.

ProviderConfig 속성에서는 제외했습니다 :)

Geun-Oh replied:
+ 환경변수에서 엔드포인트 전체를 오버라이딩 하는 경우에 대한 주석처리를 적어두기도 provider.go가 적합하다고 판단했습니다

wonchulee commented on internal/provider/provider.go:
```diff
@@ -201,6 +201,11 @@ func ProviderConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
 	}
 
+	// Set endpoint
+	if endpoint, ok := getOrFromEnv(d, "endpoint", "NCLOUD_ENDPOINT"); ok {
```
`NCLOUD_ENDPOINT` ncloud 보다 좁은 범위의 object storage에만 해당되는 값이니, 예를들면 `NCLOUD_OBS_ENDPOINT`?? 와 같은 환경 변수로 설정하면 어떨까요?

Geun-Oh replied:
좋은 의견 감사합니다. 해당 부분 나타내도록 변경해두겠습니다.

Geun-Oh committed:
chore: Add Error with Diagnostics

Geun-Oh committed:
refactor: remove endpoint from providerConfig

Geun-Oh committed:
chore: add missed packages

Geun-Oh committed:
fix: fix endpoint env to optional

Geun-Oh committed:
fix: Parse bucket info after pulling plan data

Geun-Oh committed:
refactor: Abstract BucketNameValidator

Geun-Oh committed:
Merge branch 'main' into feat/object-storage

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -0,0 +1,347 @@
+				},
+				Description: "Target bucket name",
+			},
+			"rule": schema.StringAttribute{
+				Required: true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(
+						string(awsTypes.BucketCannedACLPrivate),
+						string(awsTypes.BucketCannedACLPublicRead),
+						string(awsTypes.BucketCannedACLPublicReadWrite),
+						string(awsTypes.BucketCannedACLAuthenticatedRead),
+					),
+				},
+			},
```
이전 리뷰에서 놓친 부분입니다만, rule이 업데이트 되었을 때 동작 정의가 누락된 것으로 보여요. rule 변경시 업데이트 기능 구현 추가해주시면 좋겠습니다 (object acl도 동일)

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_data_source.go:
```diff
@@ -0,0 +1,124 @@
+			},
+			"creation_date": schema.StringAttribute{
+				Computed: true,
+				Optional: true,
```
`creation_date` attribute는 사용자 입력으로 사용되지 않고, data source를 통해 값을 읽어오는 용도로 사용되니, optional 속성은 제외해주셔도 좋겠네요!

wonchulee commented on docs/resources/object_storage_bucket.md:
```diff
@@ -0,0 +1,52 @@
+The following arguments are supported:
+
+* `bucket_name` - (Required) Bucket name to create. Bucket name must be between 3 and 63 characters long, can contain lowercase letters, numbers, periods, and hyphens. It must start and end with a letter or number, and cannot have consecutive periods.
+
```
사소한 부분이지만 attribute reference (id, creation_date) 항목이 누락된 것 같아요, 함께 추가해주시면 좋겠습니다. (bucket_acl, object_acl 리소스도 마찬가지)

wonchulee commented on docs/data-sources/object_storage_bucket.md:
```diff
@@ -0,0 +1,31 @@
+
+Prvides information about a bucket.
+
+~> **NOTE:** This resources operates in serverless environment. Does not need VPC configuration.
```
사소한 부분입니다만, `serverless environment` 의 의미가 모호할 수 있겠다는 생각이 듭니다. NCP에서는 classic, vpc 두가지 환경을 제공하고 있지만, object storage는 환경과는 무관하게 독립적으로 구성하여 사용할 수 있는 서비스로 `platform independent` 와 유사한 표현으로 설명하면 어떨까요?

Geun-Oh replied:
아아 확인했습니다. 좋은 표현 감사합니다 :)

wonchulee reviewed (commented):
Others LGTM 👍

Geun-Oh committed:
fix: Remove optional from creation_date

Geun-Oh committed:
fix: Change serverless environment => platform independent

Geun-Oh committed:
fix: change id with object_url

Geun-Oh committed:
feat: Add attribute documentation

Geun-Oh committed:
feat: Add update logic to object_acl

Geun-Oh committed:
feat: Add update logic to bucket_acl

Geun-Oh commented:
=== RUN   TestAccResourceNcloudObjectStorage_bucket_acl_basic
--- PASS: TestAccResourceNcloudObjectStorage_bucket_acl_basic (18.31s)
=== RUN   TestAccDataSourceNcloudObjectStorage_bucket_basic
--- PASS: TestAccDataSourceNcloudObjectStorage_bucket_basic (11.55s)
=== RUN   TestAccResourceNcloudObjectStorage_bucket_basic
--- PASS: TestAccResourceNcloudObjectStorage_bucket_basic (11.42s)
=== RUN   TestAccResourceNcloudObjectStorage_object_acl_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_acl_basic (26.52s)
=== RUN   TestAccResourceNcloudObjectStorage_object_copy_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_copy_basic (31.71s)
=== RUN   TestAccDataSourceNcloudObjectStorage_object_basic
--- PASS: TestAccDataSourceNcloudObjectStorage_object_basic (21.80s)
=== RUN   TestAccResourceNcloudObjectStorage_object_basic
--- PASS: TestAccResourceNcloudObjectStorage_object_basic (23.24s)
PASS
ok  	github.com/terraform-providers/terraform-provider-ncloud/internal/service/objectstorage	145.300s

wonchulee commented on internal/service/objectstorage/objectstorage_bucket_acl.go:
```diff
@@ -170,7 +170,42 @@ func (b *bucketACLResource) Read(ctx context.Context, req resource.ReadRequest,
+			return
+		}
+
+		tflog.Info(ctx, "PutObjectACL update operation response="+common.MarshalUncheckedString(response))
```
typo `PutObjectACL` -> `PutBucketACL`

Geun-Oh committed:
fix: Fix typo

Revert "fix: Fix typo"

This reverts commit 37c773743b0f01e70fd675a16f3b77592117dcf2.

fix: Fix typo

wonchulee reviewed (approved):
👍
